[DTensor][EZ] op schema comparison so that no redistribute is called

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106158


When looking at traces of TP more carefully, I found that for cases when input reshard is not needed, we also call redistribute within sharding propogation. Upon carefully checking, looks like the way we compare different op_schema is not correct.

One example can be seen in the following trace:
<img width="1146" alt="image" src="https://github.com/pytorch/pytorch/assets/6937752/7322d26f-7029-41f9-8f8c-5f27a6bb98f9">


As you can see, no collectives are called, and this redistribute is not needed.

With this change:

<img width="1491" alt="image" src="https://github.com/pytorch/pytorch/assets/6937752/eb4a971f-44c1-4d83-8671-fce94cfa926c">

fduwjj committed:
[TP] op schema comparison so that no redistribute is called

[ghstack-poisoned]

fduwjj committed:
Update on "[DTensor][EZ] op schema comparison so that no redistribute is called"



When looking at traces of TP more carefully, I found that for cases when input reshard is not needed, we also call redistribute within sharding propogation. Upon carefully checking, looks like the way we compare different op_schema is not correct.

One example can be seen in the following trace:
<img width="1146" alt="image" src="https://github.com/pytorch/pytorch/assets/6937752/7322d26f-7029-41f9-8f8c-5f27a6bb98f9">


As you can see, no collectives are called, and this redistribute is not needed.

With this change:

<img width="1491" alt="image" src="https://github.com/pytorch/pytorch/assets/6937752/eb4a971f-44c1-4d83-8671-fce94cfa926c">



[ghstack-poisoned]

fduwjj commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -159,7 +159,7 @@ def _operator_dispatch(
     assert output_sharding.schema_suggestions is not None
     suggested_input_schema = output_sharding.schema_suggestions[0]
-    needs_redistribute = suggested_input_schema is not op_schema
+    needs_redistribute = suggested_input_schema != sharding_propagator.prepare_op_schema(op_call, args, kwargs)
```
Somehow the `op_schema` got changed during sharding_prop and it's why `test_index` failed.

wanchaol replied:
re-do prepare_schema seems pretty hacky.. possibly because index op adjust the op schema. We should do some serious rewrite later.

fduwjj replied:
or maybe is it ok to disable index op test for now?

wanchaol replied:
Nah I think we still want index op to work, let's land as is and we can figure out how to remove the redundant prepare_schema later.

fduwjj committed:
Update on "[DTensor][EZ] op schema comparison so that no redistribute is called"



When looking at traces of TP more carefully, I found that for cases when input reshard is not needed, we also call redistribute within sharding propogation. Upon carefully checking, looks like the way we compare different op_schema is not correct.

One example can be seen in the following trace:
<img width="1146" alt="image" src="https://github.com/pytorch/pytorch/assets/6937752/7322d26f-7029-41f9-8f8c-5f27a6bb98f9">


As you can see, no collectives are called, and this redistribute is not needed.

With this change:

<img width="1491" alt="image" src="https://github.com/pytorch/pytorch/assets/6937752/eb4a971f-44c1-4d83-8671-fce94cfa926c">



[ghstack-poisoned]

fduwjj commented:
#pytorchbot merge -f "trunk test failure"

fduwjj commented:
@pytorchbot merge -f "trunk test failure"
